tests/python/pants_test/task/test_task.py

```diff
@@ -345,3 +401,241 @@ def test_ignore(self):
     _, vtA, was_valid = self._run_fixture(options=self._cache_ignore_options())
     self.assertFalse(was_valid)
     self.assertFalse(vtA.cacheable)
+
+class FakeTaskTest(TaskTestBase):
+  @classmethod
+  def task_type(cls):
+    return FakeTask
+
+  def context(self, **kwargs):
+    return super(TaskTestBase, self).context(**kwargs)
+
+  def _make_subtask(self, name='A', scope=GLOBAL_SCOPE, cls=FakeTask, **kwargs):
+    subclass_name = b'test_{0}_{1}_{2}'.format(cls.__name__, scope, name)
+    kwargs['options_scope'] = scope
+    return type(subclass_name, (cls,), kwargs)
+
+  def _subtask_to_fp(self, subtask, scope=GLOBAL_SCOPE, options={}):
+    context = super(TaskTestBase, self).context(for_task_types=[subtask])
+    return subtask(context, self._test_workdir).fingerprint
+
+  def _subtask_fp(self, scope=GLOBAL_SCOPE, options={}, **kwargs):
+    return self._subtask_to_fp(self._make_subtask(scope=scope, **kwargs), scope=scope, options=options)
+
+  def test_fingerprint_identity(self):
+    fpA = self._subtask_fp()
+    fpB = self._subtask_fp()
+    self.assertEqual(fpA, fpB)
+
+  def test_fingerprint_implementation_version(self):
```

Comment:
This feels like it should be broken up into separate test case methods, roughly one per comment.

I think doing that would make the differences between the cases clearer by giving them a consistent repeated structure.

Reply:
I totally agree, thanks for this comment. In the commits I've pushed since then I have tried to do this, along with renaming variables to things that actually make sense. `test_fingerprint_implementation_version_inheritance` and `test_fingerprint_options_with_scopes` are still longer than I would like, but I think they're easier to read now. Let me know if there's still more work I should do there.

Reply:
Will do!